contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,213 @@
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+	// Spec defines the behavior of a pod group.
+	// +optional
+    Spec PodGroupSpec
+
+	// Status represents the current information about a pod group.
+	// This data may not be up to date.
+	// +optional
+    Status PodGroupStatus
+}
+
+// LifecyclePolicy represents the lifecycle policy of PodGroup.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodGroup according to Event.
+    // One of "Restart", "None".
+    // Default to None.
+    // +optional
+    Action Action
+
+    // The Event recorded by scheduler; the controller takes actions
+    // according to this Event.
+    // One of "PodFailed", "Unschedulable".
+    // +optional
+    Event Event
+
+    // Timeout is the grace period for controller to take actions.
+    // Default to nil (take action immediately).
+    // +optional
+    Timeout *metav1.Duration
+}
+
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+    // NumMember defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMember int
+
+    // Policy defines the policy of PodGroup lifecycle.
+    // Default to 'Action: None, PodPhase: Failed'
+    // +optional
+    Policy []LifeCyclePolicy
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
+
+    // The number of pods which reached phase Succeeded.
+    // +optional
+    Succeeded int32
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the behavior and status of a pod group, e.g. number of pods in a group, lifecycle policy. To define which pods are member of `PodGroup`, the following field is introduced to `PodSpec`.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+    ...
+
+    // The name of PodGroup that the pod belongs to; the pod
+    // must belong to the PodGroup in the same namespace.
+    // +optional
+    GroupName string
+
+    ...
+} 
+```
+
+The `.spec.GroupName` specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
```

Comment:
> Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.

What is the difference, performance-wise, between pulling the group name out of a label, and pulling it out of a spec field?

Reply:
+1 Performance is not the main issue. Are we relaxing [rules on labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) - specifically,
"Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users, but do not directly imply semantics to the core system."

Reply:
The performance is about a selector in PodGroup; for run-to-complete workload, the status will be changed frequently; and there'll be more pending job here.

For the label, maybe annotation for alpha features as before. We do not need selector for apiserver, as scheduler will watch all pods; it need a way to build the relationship between PodGroup and Pod quickly.

BTW, the validation of `metadata.name` and `label` seems different in my mind.

Reply:
There is a big performance difference between label selectors in a PodGroup and a field in PodSpec. If we allow label selectors in PodGroups to choose pods, we would need to match the label selector of all PodGroups against any pod being scheduled to find whether the Pod is a member of the group or not. It also has the problem that a Pod may end up in multiple PodGroups. Then we would need to have some sort of priority in PodGroups to pick one.
Moreover, when any pod terminates we would again need to match the label selectors of all PodGroups against the pod to find if it was a member of a group. If so, we would need to match the selector against all running pods to determine whether the number of member has dropped below the allowed value or not.